File: uCode1/core_py/snack/schema.py
#!/usr/bin/env python3
"""
Snack schema definitions and validation - Python Implementation
"""

from typing import List, Dict, Any, Optional
from dataclasses import dataclass, field
import re


@dataclass
class SnackInputSchema:
    """Input schema for a Snack"""
    name: str
    type: str
    default: Optional[Any] = None
    required: bool = False


@dataclass
class SnackOutputSchema:
    """Output schema for a Snack"""
    name: str
    type: str


@dataclass
class SnackSchema:
    """Snack schema definition"""
    id: str
    name: str
    version: str
    kind: str
    runtime: str
    code: str
    requires: List[str] = field(default_factory=list)
    inputs: List[SnackInputSchema] = field(default_factory=list)
    outputs: List[SnackOutputSchema] = field(default_factory=list)
    tags: List[str] = field(default_factory=list)


# Valid snack kinds
VALID_SNACK_KINDS = ["script", "skill", "spark", "wasm", "snackbox", "vibe-skill"]

# Valid snack runtimes
VALID_SNACK_RUNTIMES = [
    "bash",
    "apple-script-osx", 
    "node",
    "python",
    "wasm",
    "spark-runtime",
    "vibe-skill",
    "github-spark",
]


def is_valid_semver(version: str) -> bool:
    """Check if a string is a valid semver version"""
    # Simple semver regex: major.minor.patch
    parts = version.split('.')
    if len(parts) != 3:
        return False
    
    for part in parts:
        if not part.isdigit():
            return False
    
    return True


def validate_snack_schema(schema: SnackSchema) -> None:
    """Validate a snack schema"""
    errors = []
    
    # Validate ID format
    if not schema.id:
        errors.append("Snack ID cannot be empty")

    # Validate name
    if not schema.name:
        errors.append("Snack name cannot be empty")

    # Validate version (basic semver check)
    if not is_valid_semver(schema.version):
        errors.append("Invalid version format. Use semver (e.g., 1.0.0)")

    # Validate kind
    if schema.kind not in VALID_SNACK_KINDS:
        errors.append(f"Invalid snack kind: {schema.kind}. Must be one of: {', '.join(VALID_SNACK_KINDS)}")

    # Validate runtime
    if schema.runtime not in VALID_SNACK_RUNTIMES:
        errors.append(f"Invalid runtime: {schema.runtime}. Must be one of: {', '.join(VALID_SNACK_RUNTIMES)}")

    # Validate code
    if not schema.code:
        errors.append("Snack code cannot be empty")

    # Validate inputs
    input_names = set()
    for input_schema in schema.inputs:
        if not input_schema.name:
            errors.append("Input name cannot be empty")
        if input_schema.name in input_names:
            errors.append(f"Duplicate input name: {input_schema.name}")
        input_names.add(input_schema.name)

    # Validate outputs
    output_names = set()
    for output_schema in schema.outputs:
        if not output_schema.name:
            errors.append("Output name cannot be empty")
        if output_schema.name in output_names:
            errors.append(f"Duplicate output name: {output_schema.name}")
        output_names.add(output_schema.name)

    if errors:
        raise ValueError("; ".join(errors))


# Test the schema validation
if __name__ == "__main__":
    # Test valid schema
    valid_schema = SnackSchema(
        id="P100-U899",
        name="Postie", 
        version="1.0.0",
        kind="script",
        runtime="bash",
        code="echo 'Hello'",
        requires=[],
        inputs=[],
        outputs=[],
        tags=["email"]
    )
    
    try:
        validate_snack_schema(valid_schema)
        print("✓ Valid schema passed validation")
    except ValueError as e:
        print(f"✗ Valid schema failed: {e}")
    
    # Test invalid ID
    try:
        invalid_schema = SnackSchema(
            id="",
            name="Postie",
            version="1.0.0",
            kind="script",
            runtime="bash",
            code="echo 'Hello'",
            requires=[],
            inputs=[],
            outputs=[],
            tags=[]
        )
        validate_snack_schema(invalid_schema)
        print("✗ Invalid ID should have failed")
    except ValueError as e:
        print(f"✓ Invalid ID correctly failed: {e}")
    
    # Test invalid version
    try:
        invalid_schema = SnackSchema(
            id="P100-U899",
            name="Postie",
            version="invalid",
            kind="script",
            runtime="bash",
            code="echo 'Hello'",
            requires=[],
            inputs=[],
            outputs=[],
            tags=[]
        )
        validate_snack_schema(invalid_schema)
        print("✗ Invalid version should have failed")
    except ValueError as e:
        print(f"✓ Invalid version correctly failed: {e}")
    
    # Test duplicate inputs
    try:
        input_schema = SnackInputSchema(name="mailbox", type="string", required=False)
        invalid_schema = SnackSchema(
            id="P100-U899",
            name="Postie",
            version="1.0.0",
            kind="script",
            runtime="bash",
            code="echo 'Hello'",
            requires=[],
            inputs=[input_schema, input_schema],  # Duplicate
            outputs=[],
            tags=[]
        )
        validate_snack_schema(invalid_schema)
        print("✗ Duplicate inputs should have failed")
    except ValueError as e:
        print(f"✓ Duplicate inputs correctly failed: {e}")
    
    print("All schema validation tests completed!")

File: uCode2/core/src/snack/schema.rs
//! Snack schema definitions and validation

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Snack schema definition
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SnackSchema {
    pub id: String,
    pub name: String,
    pub version: String,
    pub kind: String,
    pub runtime: String,
    pub code: String,
    pub requires: Vec<String>,
    pub inputs: Vec<SnackInputSchema>,
    pub outputs: Vec<SnackOutputSchema>,
    pub tags: Vec<String>,
}

/// Input schema for a Snack
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SnackInputSchema {
    pub name: String,
    pub r#type: String,
    pub default: Option<serde_json::Value>,
    pub required: bool,
}

/// Output schema for a Snack
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SnackOutputSchema {
    pub name: String,
    pub r#type: String,
}

/// Valid snack kinds
pub const VALID_SNACK_KINDS: [&str; 6] = ["script", "skill", "spark", "wasm", "snackbox", "vibe-skill"];

/// Valid snack runtimes
pub const VALID_SNACK_RUNTIMES: [&str; 8] = [
    "bash",
    "apple-script-osx",
    "node",
    "python",
    "wasm",
    "spark-runtime",
    "vibe-skill",
    "github-spark",
];

/// Validate a snack schema
pub fn validate_snack_schema(schema: &SnackSchema) -> Result<(), String> {
    // Validate ID format
    if schema.id.is_empty() {
        return Err("Snack ID cannot be empty".to_string());
    }

    // Validate name
    if schema.name.is_empty() {
        return Err("Snack name cannot be empty".to_string());
    }

    // Validate version (basic semver check)
    if !is_valid_semver(&schema.version) {
        return Err("Invalid version format. Use semver (e.g., 1.0.0)".to_string());
    }

    // Validate kind
    if !VALID_SNACK_KINDS.contains(&schema.kind.as_str()) {
        return Err(format!(
            "Invalid snack kind: {}. Must be one of: {}",
            schema.kind,
            VALID_SNACK_KINDS.join(", ")
        ));
    }

    // Validate runtime
    if !VALID_SNACK_RUNTIMES.contains(&schema.runtime.as_str()) {
        return Err(format!(
            "Invalid runtime: {}. Must be one of: {}",
            schema.runtime,
            VALID_SNACK_RUNTIMES.join(", ")
        ));
    }

    // Validate code
    if schema.code.is_empty() {
        return Err("Snack code cannot be empty".to_string());
    }

    // Validate inputs
    let mut input_names = HashMap::new();
    for input in &schema.inputs {
        if input.name.is_empty() {
            return Err("Input name cannot be empty".to_string());
        }
        if input_names.contains_key(&input.name) {
            return Err(format!("Duplicate input name: {}", input.name));
        }
        input_names.insert(input.name.clone(), true);
    }

    // Validate outputs
    let mut output_names = HashMap::new();
    for output in &schema.outputs {
        if output.name.is_empty() {
            return Err("Output name cannot be empty".to_string());
        }
        if output_names.contains_key(&output.name) {
            return Err(format!("Duplicate output name: {}", output.name));
        }
        output_names.insert(output.name.clone(), true);
    }

    Ok(())
}

/// Check if a string is a valid semver version
fn is_valid_semver(version: &str) -> bool {
    // Simple semver regex: major.minor.patch
    let parts: Vec<&str> = version.split('.').collect();
    if parts.len() != 3 {
        return false;
    }
    
    for part in parts {
        if part.parse::<u32>().is_err() {
            return false;
        }
    }
    
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_snack_schema() {
        let schema = SnackSchema {
            id: "P100-U899".to_string(),
            name: "Postie".to_string(),
            version: "1.0.0".to_string(),
            kind: "script".to_string(),
            runtime: "bash".to_string(),
            code: "echo 'Hello'".to_string(),
            requires: vec![],
            inputs: vec![],
            outputs: vec![],
            tags: vec!["email".to_string()],
        };

        assert!(validate_snack_schema(&schema).is_ok());
    }

    #[test]
    fn test_invalid_snack_id() {
        let mut schema = SnackSchema {
            id: "".to_string(),
            name: "Postie".to_string(),
            version: "1.0.0".to_string(),
            kind: "script".to_string(),
            runtime: "bash".to_string(),
            code: "echo 'Hello'".to_string(),
            requires: vec![],
            inputs: vec![],
            outputs: vec![],
            tags: vec![],
        };

        assert!(validate_snack_schema(&schema).is_err());
    }

    #[test]
    fn test_invalid_version() {
        let mut schema = SnackSchema {
            id: "P100-U899".to_string(),
            name: "Postie".to_string(),
            version: "invalid".to_string(),
            kind: "script".to_string(),
            runtime: "bash".to_string(),
            code: "echo 'Hello'".to_string(),
            requires: vec![],
            inputs: vec![],
            outputs: vec![],
            tags: vec![],
        };

        assert!(validate_snack_schema(&schema).is_err());
    }

    #[test]
    fn test_invalid_kind() {
        let mut schema = SnackSchema {
            id: "P100-U899".to_string(),
            name: "Postie".to_string(),
            version: "1.0.0".to_string(),
            kind: "invalid".to_string(),
            runtime: "bash".to_string(),
            code: "echo 'Hello'".to_string(),
            requires: vec![],
            inputs: vec![],
            outputs: vec![],
            tags: vec![],
        };

        assert!(validate_snack_schema(&schema).is_err());
    }

    #[test]
    fn test_duplicate_inputs() {
        let schema = SnackSchema {
            id: "P100-U899".to_string(),
            name: "Postie".to_string(),
            version: "1.0.0".to_string(),
            kind: "script".to_string(),
            runtime: "bash".to_string(),
            code: "echo 'Hello'".to_string(),
            requires: vec![],
            inputs: vec![
                SnackInputSchema {
                    name: "mailbox".to_string(),
                    r#type: "string".to_string(),
                    default: Some(serde_json::json!("inbox")),
                    required: false,
                },
                SnackInputSchema {
                    name: "mailbox".to_string(),
                    r#type: "string".to_string(),
                    default: Some(serde_json::json!("inbox")),
                    required: false,
                },
            ],
            outputs: vec![],
            tags: vec![],
        };

        assert!(validate_snack_schema(&schema).is_err());
    }
}
